File: src/policy.ts
const blockedShellTokens = ["&&", "||", ";", "|", ">", "<", "`", "$("];
const allowedCommands = new Set(["pwd", "pnpm test", "pnpm typecheck", "pnpm --version"]);
const approvableCommandPrefixes = ["pnpm install", "pnpm add", "pnpm remove"];

export type RiskLevel = "low" | "medium" | "high";

export type CommandPolicyCode =
  | "LOW_RISK_COMMAND_ALLOWED"
  | "DEPENDENCY_CHANGE_REQUIRES_APPROVAL"
  | "COMMAND_EMPTY"
  | "SHELL_OPERATOR_BLOCKED"
  | "COMMAND_NOT_ALLOWED";

export type PolicyDecision<Code extends string = string, Subject extends object = object> =
  | ({
      type: "allow";
      code: Code;
      reason: string;
    } & Subject)
  | ({
      type: "prompt";
      code: Code;
      reason: string;
      riskLevel: RiskLevel;
    } & Subject)
  | ({
      type: "forbidden";
      code: Code;
      reason: string;
    } & Subject);

export type CommandPolicyDecision = PolicyDecision<CommandPolicyCode, { command: string }>;

function normalizeCommand(command: string) {
  return command.trim().split(/\s+/).join(" ");
}

function hasShellOperator(command: string) {
  return blockedShellTokens.some((token) => command.includes(token));
}

export function getApprovableCommandPrefix(command: string) {
  return approvableCommandPrefixes.find(
    (prefix) => command === prefix || command.startsWith(`${prefix} `),
  );
}

export type RuntimeCommandPolicy = {
  allowCommandPrefix: (prefix: string) => void;
  isCommandAllowedByPrefix: (command: string) => boolean;
};

export function createRuntimeCommandPolicy(): RuntimeCommandPolicy {
  const allowedCommandPrefixes = new Set<string>();

  return {
    allowCommandPrefix(prefix) {
      allowedCommandPrefixes.add(normalizeCommand(prefix));
    },

    isCommandAllowedByPrefix(command) {
      const normalized = normalizeCommand(command);

      return Array.from(allowedCommandPrefixes).some(
        (prefix) => normalized === prefix || normalized.startsWith(`${prefix} `),
      );
    },
  };
}

export function evaluateCommandPolicy(command: string): CommandPolicyDecision {
  const trimmed = command.trim();

  if (!trimmed) {
    return {
      type: "forbidden",
      code: "COMMAND_EMPTY",
      command: "",
      reason: "Command cannot be empty.",
    };
  }

  if (hasShellOperator(trimmed)) {
    return {
      type: "forbidden",
      code: "SHELL_OPERATOR_BLOCKED",
      command: trimmed,
      reason: `Shell operator is not allowed in command: ${trimmed}`,
    };
  }

  const normalized = normalizeCommand(trimmed);

  if (allowedCommands.has(normalized)) {
    return {
      type: "allow",
      code: "LOW_RISK_COMMAND_ALLOWED",
      command: normalized,
      reason: "Known low-risk validation command.",
    };
  }

  if (getApprovableCommandPrefix(normalized)) {
    return {
      type: "prompt",
      code: "DEPENDENCY_CHANGE_REQUIRES_APPROVAL",
      command: normalized,
      reason: "Dependency command requires user approval.",
      riskLevel: "medium",
    };
  }

  return {
    type: "forbidden",
    code: "COMMAND_NOT_ALLOWED",
    command: normalized,
    reason: `Command is not allowed: ${normalized}`,
  };
}

File: internal/policy/policy.go
package policy

import (
	"strings"
)

// RiskLevel represents the risk classification of a command.
type RiskLevel string

const (
	RiskLow    RiskLevel = "low"
	RiskMedium RiskLevel = "medium"
	RiskHigh   RiskLevel = "high"
)

// CommandPolicyCode categorizes policy decisions.
type CommandPolicyCode string

const (
	CodeLowRiskCommandAllowed   CommandPolicyCode = "LOW_RISK_COMMAND_ALLOWED"
	CodeCommandRequiresApproval CommandPolicyCode = "COMMAND_REQUIRES_APPROVAL"
	CodeCommandEmpty            CommandPolicyCode = "COMMAND_EMPTY"
	CodeShellOperatorBlocked    CommandPolicyCode = "SHELL_OPERATOR_BLOCKED"
	CodeCommandNotAllowed       CommandPolicyCode = "COMMAND_NOT_ALLOWED"
)

// Decision is the result of policy evaluation.
type Decision struct {
	Type      string            `json:"type"` // "allow", "prompt", "forbidden"
	Code      CommandPolicyCode `json:"code"`
	Reason    string            `json:"reason"`
	Command   string            `json:"command"`
	RiskLevel RiskLevel         `json:"risk_level,omitempty"`
}

var (
	blockedShellTokens = []string{"&&", "||", ";", "|", ">", "<", "`", "$(", "$(("}
	allowedCommands    = map[string]struct{}{
		"pwd": {},
		// bun
		"bun --version": {},
		// npm
		"npm --version": {},
		// yarn
		"yarn --version": {},
		// pnpm
		"pnpm --version": {},
		// go
		"go version": {},
		// cargo
		"cargo --version": {},
	}
)

// Evaluate determines whether a command is allowed, requires approval, or is forbidden.
func Evaluate(command string) Decision {
	trimmed := strings.TrimSpace(command)
	if trimmed == "" {
		return Decision{Type: "forbidden", Code: CodeCommandEmpty, Reason: "Command cannot be empty.", Command: ""}
	}

	if hasShellOperator(trimmed) {
		return Decision{
			Type:    "forbidden",
			Code:    CodeShellOperatorBlocked,
			Reason:  "Shell operator is not allowed in command: " + trimmed,
			Command: trimmed,
		}
	}

	normalized := normalize(trimmed)
	if _, ok := allowedCommands[normalized]; ok {
		return Decision{Type: "allow", Code: CodeLowRiskCommandAllowed, Reason: "Known low-risk read-only command.", Command: normalized}
	}

	return Decision{
		Type:      "prompt",
		Code:      CodeCommandRequiresApproval,
		Reason:    "Command requires user approval.",
		Command:   normalized,
		RiskLevel: RiskMedium,
	}
}

// RuntimePolicy allows dynamic exact-command allowances during a session.
type RuntimePolicy struct {
	allowedCommands map[string]struct{}
}

// NewRuntimePolicy creates a fresh runtime policy store.
func NewRuntimePolicy() *RuntimePolicy {
	return &RuntimePolicy{allowedCommands: make(map[string]struct{})}
}

// AllowCommand permanently allows one exact command for this session.
func (r *RuntimePolicy) AllowCommand(command string) {
	r.allowedCommands[normalize(command)] = struct{}{}
}

// IsAllowed checks if a command exactly matches a previously allowed command.
func (r *RuntimePolicy) IsAllowed(command string) bool {
	normalized := normalize(command)
	_, ok := r.allowedCommands[normalized]
	return ok
}

func normalize(cmd string) string {
	return strings.Join(strings.Fields(cmd), " ")
}

func hasShellOperator(cmd string) bool {
	for _, token := range blockedShellTokens {
		if strings.Contains(cmd, token) {
			return true
		}
	}
	return false
}

// GetApprovableCommand returns the normalized exact command that can be
// approved for the current session.
func GetApprovableCommand(cmd string) string {
	return normalize(cmd)
}
